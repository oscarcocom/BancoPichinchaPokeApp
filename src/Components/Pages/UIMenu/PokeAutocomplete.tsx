import { FC } from "react";
import { useApiSuggestions } from "../../../Hooks/useApiSuggestions";

type Props = {
  input: string;
  setSearchInput: (paramsPokeSearch: string) => void;
  Show: boolean;
  setShow: (paramFlag: boolean) => void;
  setTableDate: (paramDte: any) => void;
  responseStatus: string;
};

type flagToReloadSuggestion = string;

export const PokeAutocomplete: FC<Props> = ({
  input,
  setSearchInput,
  setShow,
  Show,
  setTableDate,
  responseStatus,
}) => {
  const { data } =
    useApiSuggestions<flagToReloadSuggestion>(responseStatus);

  const SelectPokeOption = (selected: any) => {
    setShow(false);
    setSearchInput(selected.name);
    setTableDate({
      Id: selected.id,
      name: selected.name,
      Image: selected.image,
      attack: selected.attack,
      Defense: selected.defense,
    });
  };


  return (
    <div className="Autocomplete">
      {Show
        ? data
            .filter((fil: any) =>  {
              if (
                fil.name.toLowerCase().indexOf(input.toLowerCase()) > -1 ||
                fil.id === input
              )
                return fil;
            })
            .map((items: any) => (
              <div
                key={items.id}
                className="Option"
                onClick={() => SelectPokeOption(items)}
              >
                <span className="textItem">{items.name}</span>
                <span className="imageItem">
                  <img src={items.image} alt="Pokemon" />
                </span>
              </div>
            ))
        : null}
    </div>
  );
};
