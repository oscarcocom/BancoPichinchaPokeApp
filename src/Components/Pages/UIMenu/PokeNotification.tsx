import React,{FC} from "react";
import { PokePopMessage } from '../../../Interfaces/Pokemon-Api-post';

interface Props{
    Visible:boolean,
    Message:string,
    setPokePopMessage:(paramFlag: any) => void;
}

export const PokeNotification:FC<Props> = ({Visible,Message, setPokePopMessage}) => {
  return (
    <div>
  {
    Visible?
    <div className=" ShowPokeNotification">
        {/*
              IMPORTANTE !!
              Siempre pensando en las reglas de NO framework de estilos o componentes prefabricados, en esta parte hice uso de una libreria de animacion, esta libreria solo anima el texto y divs y no es un auxiliar CSS en el maquetado, el cual fue hecho 100% css a mano, decidi usarlo como propuesta propia ya que me parecio una buena idea de uso, saludos ! :)
              All along thinked in the rules of never use framework of styles for original design so I decided use in this part of my code a library to do text animated and not as css helper ! :) */}
        <div className=" animate__animated animate__swing popup">
          <h2>Poke Notificaci√≥n!</h2>
          <a className="close" 
             onClick={()=>setPokePopMessage({
              Visible:false,
              Message:"Operacion hecha"
             })}
             href="#">
            &times;
          </a>
          <div className="content">
          {Message}
          </div>
        </div>
      </div>:null
      
  }
      
    </div>
  );
};
