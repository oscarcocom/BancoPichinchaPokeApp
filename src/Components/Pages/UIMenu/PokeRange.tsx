import styled from 'styled-components'

type propsRange= any;
const PokeRange = styled.input.attrs({ type: 'range' })`
  -webkit-appearance: none;
  -moz-appearance: none;
  outline: 0;
  height: 7px;
  width: 13rem;
  border-radius: 40px;
  background: ${(props:propsRange) =>
    `linear-gradient(to right, rgb(89, 31, 196) 0%, #6f00ff ${props.value}%, #977fcc9e ${props.value}%, #977fcc85  0%);`};


  ::-webkit-slider-thumb {
    -webkit-appearance: none;
    width: 20px;
    height: 20px;
    background-image: radial-gradient(circle, #4934b0 40%, #162cd4);
    border-radius: 50%;
    
  }

  ::-moz-range-thumb {
    width: 24px;
    height: 18px;
    -moz-appearance: none;
    background-image: radial-gradient(circle, #282895 40%, #271696 45%);
    border-radius: 50%;
   
  }
`;

export default PokeRange;