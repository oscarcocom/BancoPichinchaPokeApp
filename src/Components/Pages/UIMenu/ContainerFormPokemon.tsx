import {FC} from 'react'
import { RiCloseLine } from "react-icons/ri";
import { TbLoader } from "react-icons/tb";
import PokeRange from './PokeRange';
import {MdOutlineSave } from "react-icons/md";
import { TopLevelPost, TopLevelPut } from '../../../Interfaces/Pokemon-Api-post';
import pokeApi from '../../../Api/getApi';

interface ProposContainer{
    HandleMethod:string,
    InputName:any,
    Name:string,
    HandleFormCrud:(params:any)=>void,
    InputImagen:any,
    Imagen:string,
    ApiLoader:any;
    setAtaque:(param:any)=>void,
    Ataque:any,
    setDefensa:(param:any)=>void,
    Defensa:any,
    setApiLoader:(param:any)=>void,
    setPokePopMessage:(param:any)=>void;
    IdPomenon:number,
    setApiMethod:(param:any)=>void;
    NewPokenon:(param:any)=>void;
  
}
export const ContainerFormPokemon:FC<ProposContainer>=(
    {
        HandleMethod="Post",
        InputName,
        Name,
        HandleFormCrud,
        InputImagen,
        Imagen,
        ApiLoader,
        Ataque,
        setAtaque,
        Defensa,
        setDefensa,
        setApiLoader,
        setPokePopMessage,
        IdPomenon,
        setApiMethod,
        NewPokenon
                                        
    
    }) => {




        const HandleCrudPost = (e: React.FormEvent<HTMLFormElement>) => {
            e.preventDefault();
        
            setApiLoader(true);
        
            if ([Name, Imagen].includes("")) {
              setPokePopMessage({
                Visible: true,
                Message: "Hay campos Input vacio",
              });
              setApiLoader(false);
              return;
            }
            if (Ataque <= 0) {
              setPokePopMessage({
                Visible: true,
                Message: "El pokemon debe tener un ataque arriba de 0",
              });
              setApiLoader(false);
              return;
            }
            if (Defensa <= 0) {
              setPokePopMessage({
                Visible: true,
                Message: "El pokemon debe tener una Defensa arriba de 0",
              });
              setApiLoader(false);
              return;
            }
        
            const Load = async () =>
              await pokeApi
                .post<TopLevelPost>(`?idAuthor=1`, {
                  idAuthor: 1,
                  name: Name,
                  image: Imagen,
                  attack: Ataque,
                  defense: Defensa,
                  hp: 100,
                  type: "Dragon",
                })
                .then((pokeData) => {

                  setApiMethod((f:any)=>({
                    ...f,
                    responseStatus:"POST ALL DONE !"
                  }));
                  console.log(pokeData);
                  setPokePopMessage({
                    Visible: true,
                    Message: `Operaci贸n POST exitosa con estatus de operaci贸n: ${pokeData.status} ${pokeData.statusText}, ya puede buscar su nuevo pokemon ${Name}  en la Api texteando en el input de busqueda `,
                  });
                

               

                })
                .catch((error) => {
                  // console.error(error)
                  setPokePopMessage({
                    Visible: true,
                    Message: `Error sucedido ${error}`,
                  });
                })
                .finally(() => {
               
               
                  setApiLoader(false)
                });
        
            Load();
                // RESETEAMOS LA DATA SIN PERDER EL METODO REST PARA ACTUALIZAR LAS SUGERENCIAS
                //I RESET STATE FLAG TO GET NEW DATA
            setApiMethod((f:any)=>({
              ...f,
              responseStatus:"OK"
            }));  
          };
          const HandleCrudPut =  (e: React.FormEvent<HTMLFormElement>) => {
            e.preventDefault();
        
            setApiLoader(true);
        
            if ([Name, Imagen].includes("")) {
              setPokePopMessage({
                Visible: true,
                Message: "Hay campos Input vacio",
              });
              setApiLoader(false);
              return;
            }
            if (Ataque <= 0) {
              setPokePopMessage({
                Visible: true,
                Message: "El pokemon debe tener un ataque arriba de 0",
              });
              setApiLoader(false);
              return;
            }
            if (Defensa <= 0) {
              setPokePopMessage({
                Visible: true,
                Message: "El pokemon debe tener una Defensa arriba de 0",
              });
              setApiLoader(false);
              return;
            }
        
            const Load = async () =>
              await pokeApi
                .put<TopLevelPut>(`${IdPomenon}`, {
                  name: Name,
                  image: Imagen,
                  attack: Ataque,
                  defense: Defensa,
                  hp: 100,
                  type: "Dragon",
                  idAuthor: 1
                })
                .then((pokeData) => {
                  console.log(pokeData);
                  

                  setApiMethod((f:any)=>({
                    ...f,
                    responseStatus:"PUT ALL DONE !"
                  }));

                  setPokePopMessage({
                    Visible: true,
                    Message: `Operaci贸n PUT exitosa con estatus de operaci贸n: ${pokeData.status} ${pokeData.statusText},ya puede verificar su actualizacion para el pokemon ${Name} en la Api texteando en el input de busqueda  `,
                  });
                })
                .catch((error) => {
                  // console.error(error)
                  setPokePopMessage({
                    Visible: true,
                    Message: `Error sucedido ${error}`,
                  });
                })
                .finally(() =>{
                  // RESETEAMOS LA DATA SIN PERDER EL METODO REST
                
                  setApiLoader(false)
                  
                  });
        
            Load();
               // RESETEAMOS LA DATA SIN PERDER EL METODO REST PARA ACTUALIZAR LAS SUGERENCIAS
                //I RESET STATE FLAG TO GET NEW DATA
                setApiMethod((f:any)=>({
                  ...f,
                  responseStatus:"OK"
                }));  
          };
        

  return (
   <>
   
       <div className="fila">
            <div className="titleAddBox columna-s-12 columna-6">
              
                {
                    HandleMethod==="Post"?
                   <h4>Nuevo Pokemon</h4> 
                    : 
                    <h4 >Ahora se actualiza al pokemon : <p style={{ display:"inline-block", color:"purple"}}>{Name}</p></h4>
                }
                
            </div>
          </div>
          <div className="fila">
            <form onSubmit={HandleMethod==="Post"?HandleCrudPost:HandleCrudPut } className="FormAddPokemon">
              <div  className="FormContent">
                {/* inputs */}
                <div className="inputsBox columna-s-12 columna-8">
                  {/* Nombre */}

                  <div className="NombreAdd  columna-s-12 columnna-6">
                    <label htmlFor="Nombre">Nombre:</label>
                    <input
                      ref={InputName}
                      value={Name}
                      name="Name"
                      placeholder="Nombre"
                      onChange={HandleFormCrud}
                    />
                  </div>
                  <div className="ImgAdd  columna-s-12 columnna-6">
                    <label htmlFor="Nombre">Imagen:</label>
                    <input
                      ref={InputImagen}
                      value={Imagen}
                      name="Imagen"
                      placeholder="Url"
                      onChange={HandleFormCrud}
                    />
                  </div>

                  {/* url */}
                </div>

                {/* ranges */}
                <div className="RangesBox columna-s-12 columna-8">
                  <div className="ataque">
                    <label htmlFor="">Ataque: {Ataque} </label>
                    <PokeRange
                      value={Ataque}
                      onChange={(e: any) => setAtaque(e.target.value)}
                    />{" "}
                    <label htmlFor="">100</label>
                  </div>

                  <div className="defensa ">
                    <label htmlFor="">Defensa: {Defensa} </label>
                    <PokeRange
                      value={Defensa}
                      onChange={(e: any) => setDefensa(e.target.value)}
                    />{" "}
                    <label htmlFor="">100</label>
                  </div>
                </div>
              </div>
              <div className="FormContent">
                {/* Footer botones */}
                <div className="Footer fila">
                  <div className="ButtonBoxFooter columna-8 columna-s-8">
                 {
                 
                ApiLoader?
                    <button disabled className="LoaderBotton">
                      <TbLoader className="icons" />
                      <span>Cargando</span>
                    </button>
                    : 
                    <button>
                    <MdOutlineSave  className="icons" />
                    <span>Guardar</span>
                  </button>
                    }
                    <button onClick={NewPokenon} >
                      <RiCloseLine className="icons" />
                      <span >Cancelar</span>
                    </button>
                  </div>
                </div>
              </div>
            </form>
          </div>
   </>
  )
}
