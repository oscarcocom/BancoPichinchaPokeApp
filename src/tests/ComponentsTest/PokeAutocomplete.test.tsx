import { render, renderHook } from "@testing-library/react"
import { PokeAutocomplete } from '../../Components/Pages/UIMenu/PokeAutocomplete';
import { useState } from 'react';

//APLICACION TDD 
// RED AL NO MATCH CON SCREEN
//GREEN AL MATCH
// REFACTOR DONE

describe('Prueba de snapshoot al componente <PokeAutocomplete>', () => {

   const {result}= renderHook(()=>{
        const [SearchInput, setSearchInput] = useState("");
        const [Show, setShow] = useState(false);
        const [{Id, name, Image, attack, Defense }, setTableDate] = useState({
            Id:0,
            name: "Waiting...",
            Image: "",
            attack: "Waiting...",
            Defense: "Waiting...",
          });
          const [{CurrentMethod,responseStatus}, setApiMethod] = useState({
            CurrentMethod: "Post",
            responseStatus:"OK"
          });
        

        return{
            SearchInput,
            setSearchInput,
            Show,
            setShow,
            setTableDate,
            responseStatus
        }
    })


test('Mostrar sugerencias del API', () => { 



  const{container}=render(<PokeAutocomplete 
        input={ result.current.SearchInput}
        setSearchInput={  result.current.setSearchInput}
        setShow={   result.current.setShow}
        Show={   result.current.Show}
        setTableDate={   result.current.setTableDate}
        responseStatus={  result.current.responseStatus}
        />)


       expect(container).toMatchSnapshot() 
 })

 })