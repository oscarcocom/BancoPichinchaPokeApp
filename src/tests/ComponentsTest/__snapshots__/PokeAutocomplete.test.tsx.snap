// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba de snapshoot al componente <PokeAutocomplete> Mostrar sugerencias del API 1`] = `
<div>
  <div
    class="Autocomplete"
  />
</div>
`;
